src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,212 @@
+  };
+
+  return gcs_client_.Actors().AsyncSubscribe(actor_notification_callback, nullptr);
+}
+
+void CoreWorkerDirectActorTaskSubmitter::HandleActorAlive(const ActorID &actor_id,
+                                                          std::string ip_address,
+                                                          int port) {
+  std::unique_ptr<rpc::DirectActorClient> grpc_client(
+      new rpc::DirectActorClient(ip_address, port, client_call_manager_));
+  // replace old rpc client if it exists.
+  rpc_clients_[actor_id] = std::move(grpc_client);
+
+  // Submit all pending requests.
+  auto &client = rpc_clients_[actor_id];
+  auto &requests = pending_requests_[actor_id];
+  while (!requests.empty()) {
+    const auto &request = *requests.front();
+    auto status =
+        PushTask(*client, *request.request_, request.task_id_, request.num_returns_);
+    requests.pop_front();
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::PushTask(rpc::DirectActorClient &client,
+                                                    const rpc::PushTaskRequest &request,
+                                                    const TaskID &task_id,
+                                                    int num_returns) {
+  auto status = client.PushTask(
+      request,
+      [this, task_id, num_returns](Status status, const rpc::PushTaskReply &reply) {
+        if (!status.ok()) {
+          TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+          return;
+        }
+        for (int i = 0; i < reply.return_objects_size(); i++) {
+          const auto &return_object = reply.return_objects(i);
+          ObjectID object_id = ObjectID::FromBinary(return_object.object_id());
+          std::shared_ptr<LocalMemoryBuffer> data_buffer;
+          if (return_object.data().size() > 0) {
+            data_buffer = std::make_shared<LocalMemoryBuffer>(
+                const_cast<uint8_t *>(
+                    reinterpret_cast<const uint8_t *>(return_object.data().data())),
+                return_object.data().size());
+          }
+          std::shared_ptr<LocalMemoryBuffer> metadata_buffer;
+          if (return_object.metadata().size() > 0) {
+            metadata_buffer = std::make_shared<LocalMemoryBuffer>(
+                const_cast<uint8_t *>(
+                    reinterpret_cast<const uint8_t *>(return_object.metadata().data())),
+                return_object.metadata().size());
+          }
+          store_provider_->Put(RayObject(data_buffer, metadata_buffer), object_id);
+        }
+      });
+  return status;
+}
+
+void CoreWorkerDirectActorTaskSubmitter::TreatTaskAsFailed(
+    const TaskID &task_id, int num_returns, const rpc::ErrorType &error_type) {
+  for (int i = 0; i < num_returns; i++) {
+    const auto object_id = ObjectID::ForTaskReturn(task_id, i + 1);
+    std::string meta = std::to_string(static_cast<int>(error_type));
+    auto data = const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(meta.data()));
```

Comment:
```suggestion
    auto metadata = const_cast<uint8_t *>(reinterpret_cast<const uint8_t *>(meta.data()));
```
I thought it's "data of the object" initially. maybe rename to `metadata` to avoid confusion.